### PR TITLE
feat(api-client): Add new MLS endpoints (FS-652)

### PR DESCRIPTION
**Pull request by thisisamir98**
## Pull Request Checklist

- [X] My code is covered by tests
- [X] I will [merge the PR as breaking change](https://github.com/wireapp/wire-web-packages/wiki/Releases#create-major-release), if the API contract changes

### CODE REVIEW

File: packages/api-client/src/conversation/ConversationAPI.ts [hunk 1]
```diff
@@ -791,6 +793,25 @@ export class ConversationAPI {
...
+  /**
+   * Post an encrypted message to a conversation.
+   * @param messageData Mls message payload in TLS format. Please refer to the MLS specification for details.
+   * @see https://messaginglayersecurity.rocks/mls-protocol/draft-ietf-mls-protocol.html#name-message-framing
+   * @see https://staging-nginz-https.zinfra.io/swagger-ui/#!/conversations/postOtrMessage
+   */
+  public async postMlsMessage(messageData: Uint8Array): Promise<MlsEvent> {
+    const config: AxiosRequestConfig = {
+      data: {
+        // TODO: use core crypto here to encrypt the message data
```
Comment (atomrc) [inline packages/api-client/src/conversation/ConversationAPI.ts:805]: Side note: The core-crypto lib will never be known from the api-client, it's going to live in the `cryptoRepo` of the `core` ;)

File: packages/api-client/src/conversation/data/MlsEventData.ts [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+  conversation: string;
+  data: MlsEventData;
+  from: string;
+  qualified_conversation: QualifiedId;
+  qualified_from: QualifiedId;
+  time: string;
+  type: string;
+}
+
+export interface MlsEventData {
```
Comment (atomrc) [inline packages/api-client/src/conversation/data/MlsEventData.ts:35]: Sweeeettt, thanks for adding the clean types :pray:

File: packages/api-client/src/conversation/ConversationAPI.ts (hunk 1 above)
Comment (thisisamir98) [reply packages/api-client/src/conversation/ConversationAPI.ts]: I removed this comment

File: packages/api-client/src/client/ClientAPI.ts [hunk 3]
```diff
@@ -137,4 +148,36 @@ export class ClientAPI {
...
+  /**
+   * Claim one key package for each client of the given user
+   * @param  {string} userId The user to claim the key packages for
+   * @param {string} userDomain The domain of the user
+   * @param  {string} skipOwn Do not claim a key package for the given own client
+   * @see https://staging-nginz-https.zinfra.io/api/swagger-ui/#/default/post_mls_key_packages_claim__user_domain___user_
+   */
+  public async claimMLSKeyPackages(userId: string, userDomain: string, skipOwn: string): Promise<ClaimedKeyPackages> {
+    const config: AxiosRequestConfig = {
+      data: {skip_own: skipOwn},
```
Comment (tlebon) [inline packages/api-client/src/client/ClientAPI.ts:161]: looks like this should be included as a query, not as a request body

Comment (thisisamir98) [reply packages/api-client/src/client/ClientAPI.ts]: nice catch man, fixed in https://github.com/wireapp/wire-web-packages/pull/4292/commits/c9819bd0fdac6430a8fbce45e57f0d10a8a14b9d

File: packages/api-client/src/conversation/ConversationAPI.ts [hunk 4]
```diff
@@ -792,6 +794,43 @@ export class ConversationAPI {
...
+  /**
+   * Post an encrypted message to a conversation.
+   * @param messageData Mls message payload in TLS format. Please refer to the MLS specification for details.
+   * @see https://messaginglayersecurity.rocks/mls-protocol/draft-ietf-mls-protocol.html#name-message-framing
+   * @see https://staging-nginz-https.zinfra.io/swagger-ui/#!/conversations/postOtrMessage
+   */
+  public async postMlsMessage(messageData: Uint8Array): Promise<MlsEvent> {
+    const config: AxiosRequestConfig = {
+      data: {
+        description: messageData,
```
Comment (tlebon) [inline packages/api-client/src/conversation/ConversationAPI.ts:806]: i think theres something i dont fully understand but based on the swagger API the message body looks like a plaintext string?

File: packages/api-client/src/conversation/ConversationAPI.ts [hunk 5]
```diff
@@ -792,6 +794,43 @@ export class ConversationAPI {
...
+  /**
+   * Post the welcome encrypted message to a conversation.
+   * @param messageData Mls welocome message payload in TLS format. Please refer to the MLS specification for details.
+   * @see https://messaginglayersecurity.rocks/mls-protocol/draft-ietf-mls-protocol.html#name-message-framing
+   * @see https://staging-nginz-https.zinfra.io/api/swagger-ui/#/default/post_mls_welcome
+   */
+  public async postMlsWelcomeMessage(messageData: Uint8Array): Promise<void> {
+    const config: AxiosRequestConfig = {
+      data: {
+        description: messageData,
```
Comment (tlebon) [inline packages/api-client/src/conversation/ConversationAPI.ts:825]: relevant here as well

File: packages/api-client/src/conversation/ConversationAPI.ts (hunk 4 above)
Comment (thisisamir98) [reply packages/api-client/src/conversation/ConversationAPI.ts]: I was in doubt too but I did it this way cause I looked at the Model in swagger but think you're maybe right it makes more sense 
https://github.com/wireapp/wire-web-packages/pull/4292/commits/4d5200568eea097cefd3ef945431ba93abf55463

File: packages/api-client/src/conversation/ConversationAPI.ts (hunk 5 above)
Comment (thisisamir98) [reply packages/api-client/src/conversation/ConversationAPI.ts]: https://github.com/wireapp/wire-web-packages/pull/4292/commits/4d5200568eea097cefd3ef945431ba93abf55463

File: packages/api-client/src/conversation/ConversationAPI.ts [hunk 6]
```diff
@@ -792,6 +794,39 @@ export class ConversationAPI {
     return response.data;
   }
 
+  /**
+   * Post an encrypted message to a conversation.
+   * @param messageData Mls message payload in TLS format. Please refer to the MLS specification for details.
+   * @see https://messaginglayersecurity.rocks/mls-protocol/draft-ietf-mls-protocol.html#name-message-framing
+   * @see https://staging-nginz-https.zinfra.io/swagger-ui/#!/conversations/postOtrMessage
```
Comment (tlebon) [inline packages/api-client/src/conversation/ConversationAPI.ts:801]: reference here is outdated, points to Otr

File: packages/api-client/src/conversation/ConversationAPI.ts [hunk 7]
```diff
@@ -792,6 +794,39 @@ export class ConversationAPI {
...
+   */
+  public async postMlsMessage(messageData: Uint8Array): Promise<MlsEvent> {
```
Comment (tlebon) [inline packages/api-client/src/conversation/ConversationAPI.ts:803]: is there a possibility the data type could also be a string? it was the case for Otr messages, not sure how its working for Mls @atomrc

Comment (atomrc) [reply packages/api-client/src/conversation/ConversationAPI.ts:803]: For OTR messages, that was no string, it was a Protobuf that contained a `string`. 
But it gets encoded into a Uint8Array right before sending :) (see https://github.com/wireapp/wire-web-packages/blob/main/packages/api-client/src/conversation/ConversationAPI.ts#L673)

File: packages/api-client/src/client/ClientAPI.ts [hunk 8]
```diff
@@ -125,8 +135,9 @@ export class ClientAPI {
 
   /**
    * Will upload keypackages for an MLS capable client
-   * see https://staging-nginz-https.zinfra.io/api/swagger-ui/#/default/post_mls_key_packages_self__client_
-   * @param  {string} clientId
+   * @see https://staging-nginz-https.zinfra.io/api/swagger-ui/#/default/post_mls_key_packages_self__client_
+   * @param  {string} clientId The client to upload the key packages for
+   * @param  {Array.<string>} keyPackages The key packages to upload
```
Comment (atomrc) [inline packages/api-client/src/client/ClientAPI.ts:140]:
```suggestion
   * @param {string} clientId The client to upload the key packages for
   * @param {string[]} keyPackages The key packages to upload
```

I am curious where this `Array.<string>` comes from, is the "." a typo?

File: packages/api-client/src/client/ClientAPI.ts (hunk 3 above)
Comment (atomrc) [reply packages/api-client/src/client/ClientAPI.ts]: I wonder why this endpoint is actually a `POST` endpoint :thinking:

File: packages/api-client/src/client/ClientAPI.ts (hunk 8 above)
Comment (tlebon) [reply packages/api-client/src/client/ClientAPI.ts:140]: its another option for typing in jsdoc, but I agree that synchronising with typescript types is preferable
https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#types-1

File: packages/api-client/src/conversation/ConversationAPI.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (tlebon) [reply packages/api-client/src/conversation/ConversationAPI.ts:801]: 😅 https://github.com/wireapp/wire-web-packages/pull/4292#discussion_r900162753

**Review by atomrc — APPROVED** [review]

**Pull request merged**